sst/SimengMemInterface.cc

```diff
@@ -0,0 +1,279 @@
+    /* 
+        Check if address is greater than max memory address or overflows.
+        This often happens on wrongly speculated branches leading to 
+        large values. In this case we queue an empty register value 
+        which signals an exception. However, wrongly speculated branches
+        lead to a pipeline flush after which execution continues.
+    */
+    if (addrEnd > max_addr_memory || unsignedOverflow_(addrStart, size)) {
+        completed_read_requests.push_back({target, RegisterValue(), requestId});
+        return;
+    }
+
+    AggregateReadRequest* aggrReq = new AggregateReadRequest(target, requestId);
+    std::vector<StandardMem::Request*> requests = makeSSTRequests<AggregateReadRequest>(aggrReq, addrStart, addrEnd, size);
+    for (StandardMem::Request* req : requests) {
+        mem->send(req);
+    }
+}
+
+void SimengMemInterface::requestWrite(const MemoryAccessTarget& target, const RegisterValue& data) {
+    uint64_t addrStart = target.address;
+    uint64_t size = unsigned(target.size);
+    uint64_t addrEnd = addrStart + size - 1;
+
+    AggregateWriteRequest* aggrReq = new AggregateWriteRequest(target, data);
+    std::vector<StandardMem::Request*> requests = makeSSTRequests<AggregateWriteRequest>(aggrReq, addrStart, addrEnd, size);
+    for (StandardMem::Request* req : requests) {
+        mem->send(req);
+    }
+}
+
+void SimengMemInterface::tick() {
+    tickCounter++;
+}
+
+void SimengMemInterface::clearCompletedReads() {
+    completed_read_requests.clear();
+}
+
+bool SimengMemInterface::hasPendingRequests() const {
+    return completed_read_requests.size() > 0;
+};
+
+const span<MemoryReadResult> SimengMemInterface::getCompletedReads() const {
+    return {const_cast<MemoryReadResult*>(completed_read_requests.data()),
+        completed_read_requests.size()};
+};
+
+void SimengMemInterface::aggregatedReadResponses(AggregateReadRequest* aggrReq) {
+
+    if (aggrReq->aggregateCount != 0) return;
+    std::vector<uint8_t> mergedData;
+    // Loop through the ordered map and merge the data in order inside the mergedData vector.
+    // Also remove entries from the aggregation_map as we loop through each SST Request id.
+    for (auto itr = aggrReq->response_map.begin(); itr != aggrReq->response_map.end(); itr++)
+    {
+        mergedData.insert(mergedData.end(), itr->second.begin(), itr->second.end());
+        aggregation_map.erase(itr->first);
+    }
+    // Send the completed read request back to SimEng via the completed_read_requests
+    // queue.
+    const char* char_data = reinterpret_cast<const char*>(&mergedData[0]);
+    completed_read_requests.push_back({
+        aggrReq->target,
+        RegisterValue(char_data, uint16_t(unsigned(aggrReq->target.size))),
+        aggrReq->id
+    });
+
+    // Cleanup
+    aggrReq->response_map.clear();
+    delete aggrReq;
+}
+
+void SimengMemInterface::SimengMemHandlers::handle(StandardMem::WriteResp* rsp) {
+    delete rsp;
```

Comment:
Are we always assuming the write request was successful? Can an SST write request fail? If so, we should be able to handle it and act accordingly

Reply:
If a request fails SST breaks and the program terminates.

Reply:
Is that expected behaviour (SST just breaks)? Does it safely break (i.e. exit with an error code or eqv)?

Reply:
It does break safely with an error code. There is nothing we can do to catch it as it is an internal error are we are running simeng inside SST.